drivers/video/video_stm32_dcmi.c

```diff
@@ -0,0 +1,511 @@
+	struct stream dma;
+	struct video_format fmt;
+	struct k_fifo fifo_in;
+	struct k_fifo fifo_out;
+	uint32_t pixel_format;
+	uint32_t height;
+	uint32_t width;
+	uint32_t pitch;
+	uint8_t *buffer;
+};
+
+struct video_stm32_dcmi_config {
+	struct stm32_pclken pclken;
+	irq_config_func_t irq_config;
+	const struct pinctrl_dev_config *pctrl;
+	const struct device *sensor_dev;
+};
+
+static inline unsigned int video_pix_fmt_bpp(uint32_t pixelformat)
+{
+	switch (pixelformat) {
+	case VIDEO_PIX_FMT_BGGR8:
+	case VIDEO_PIX_FMT_GBRG8:
+	case VIDEO_PIX_FMT_GRBG8:
+	case VIDEO_PIX_FMT_RGGB8:
+		return 1;
+	case VIDEO_PIX_FMT_RGB565:
+	case VIDEO_PIX_FMT_YUYV:
+		return 2;
+	default:
+		return 0;
+	}
+}
+
+void HAL_DCMI_ErrorCallback(DCMI_HandleTypeDef *hdcmi)
+{
+	LOG_WRN("%s", __func__);
+}
+
+void HAL_DCMI_FrameEventCallback(DCMI_HandleTypeDef *hdcmi)
+{
+	struct video_stm32_dcmi_data *dev_data =
+			CONTAINER_OF(hdcmi, struct video_stm32_dcmi_data, hdcmi);
+	struct video_buffer *vbuf;
+
+	HAL_DCMI_Suspend(hdcmi);
+
+	vbuf = k_fifo_get(&dev_data->fifo_in, K_NO_WAIT);
+
+	if (vbuf == NULL) {
+		LOG_DBG("Failed to get buffer from fifo");
+		goto resume;
+	}
+
+	vbuf->timestamp = k_uptime_get_32();
+	memcpy(vbuf->buffer, dev_data->buffer, vbuf->bytesused);
+
+	k_fifo_put(&dev_data->fifo_out, vbuf);
+
+resume:
+	HAL_DCMI_Resume(hdcmi);
+}
+
+static void stm32_dcmi_isr(const struct device *dev)
+{
+	struct video_stm32_dcmi_data *data = dev->data;
+
+	HAL_DCMI_IRQHandler(&data->hdcmi);
+}
+
+static void dmci_dma_callback(const struct device *dev, void *arg,
+			 uint32_t channel, int status)
+{
+	DMA_HandleTypeDef *hdma = arg;
+
+	ARG_UNUSED(dev);
+
+	if (status < 0) {
+		LOG_ERR("DMA callback error with channel %d.", channel);
+	}
+
+	HAL_DMA_IRQHandler(hdma);
+}
+
+void HAL_DMA_ErrorCallback(DMA_HandleTypeDef *hdma)
+{
+	LOG_WRN("%s", __func__);
+}
+
+static int stm32_dma_init(const struct device *dev)
+{
+	struct video_stm32_dcmi_data *data = dev->data;
+	int ret;
+
+	/* Check if the DMA device is ready */
+	if ((data->dma.dma_dev != NULL) &&
+		!device_is_ready(data->dma.dma_dev)) {
+		LOG_ERR("%s DMA device not ready", data->dma.dma_dev->name);
+		return -ENODEV;
+	}
+
+	/*
+	 * DMA configuration
+	 * Due to use of QSPI HAL API in current driver,
+	 * both HAL and Zephyr DMA drivers should be configured.
+	 * The required configuration for Zephyr DMA driver should only provide
+	 * the minimum information to inform the DMA slot will be in used and
+	 * how to route callbacks.
+	 */
+	struct dma_config dma_cfg = data->dma.cfg;
+	static DMA_HandleTypeDef hdma;
+
+	/* Proceed to the minimum Zephyr DMA driver init */
+	dma_cfg.user_data = &hdma;
+	/* HACK: This field is used to inform driver that it is overridden */
+	dma_cfg.linked_channel = STM32_DMA_HAL_OVERRIDE;
+	/* Because of the STREAM OFFSET, the DMA channel given here is from 1 - 8 */
+	ret = dma_config(data->dma.dma_dev,
+			 (data->dma.channel + STM32_DMA_STREAM_OFFSET), &dma_cfg);
```

Comment:
no need for the parenthesis around `data->dma.channel + STM32_DMA_STREAM_OFFSET`, if you think it helps clarity you can use an intermediate variable but I would drop them

Reply:
Removed